flink-external-resources/flink-external-resource-gpu/src/main/java/org/apache/flink/externalresource/gpu/GPUDriver.java

```diff
@@ -0,0 +1,143 @@
+
+/**
+ * Driver takes the responsibility to discover GPU resources and provide the GPU resource information.
+ * It get the GPU information by executing user-defined discovery script.
+ */
+class GPUDriver implements ExternalResourceDriver {
+
+	private static final Logger LOG = LoggerFactory.getLogger(GPUDriver.class);
+
+	private static final long DISCOVERY_SCRIPT_TIMEOUT_MS = 10000;
+
+	@VisibleForTesting
+	static final ConfigOption<String> DISCOVERY_SCRIPT_PATH =
+		key("discovery-script.path")
+			.stringType()
+			.defaultValue(String.format("%s/external-resource-gpu/nvidia-gpu-discovery.sh", ConfigConstants.DEFAULT_FLINK_PLUGINS_DIRS));
+
+	@VisibleForTesting
+	static final ConfigOption<String> DISCOVERY_SCRIPT_ARG =
+		key("discovery-script.args")
+			.stringType()
+			.noDefaultValue();
+
+	private final File discoveryScriptFile;
+	private final String args;
+
+	GPUDriver(Configuration config) throws Exception {
+		final String discoveryScriptPathStr = config.getString(DISCOVERY_SCRIPT_PATH);
+		if (StringUtils.isNullOrWhitespaceOnly(discoveryScriptPathStr)) {
+			throw new IllegalConfigurationException(
+				String.format("GPU discovery script ('%s') is not configured.", ExternalResourceOptions.genericKeyWithSuffix(DISCOVERY_SCRIPT_PATH.key())));
+		}
+
+		Path discoveryScriptPath = Paths.get(discoveryScriptPathStr);
+		if (!discoveryScriptPath.isAbsolute()) {
+			discoveryScriptPath = Paths.get(System.getenv().getOrDefault(ConfigConstants.ENV_FLINK_HOME_DIR, "."), discoveryScriptPathStr);
+		}
+		discoveryScriptFile = discoveryScriptPath.toFile();
+
+		if (!discoveryScriptFile.exists()) {
+			throw new FileNotFoundException(String.format("The gpu discovery script does not exist in path %s.", discoveryScriptFile.getAbsolutePath()));
+		}
+		if (!discoveryScriptFile.canExecute()) {
+			throw new FlinkException(String.format("The discovery script %s is not executable.", discoveryScriptFile.getAbsolutePath()));
+		}
+
+		args = config.getString(DISCOVERY_SCRIPT_ARG);
+	}
+
+	@Override
+	public Set<GPUInfo> retrieveResourceInfo(long gpuAmount) throws Exception {
+		Preconditions.checkArgument(gpuAmount > 0, "The gpuAmount should be positive while finding " + gpuAmount);
+
+		final Set<GPUInfo> gpuResources = new HashSet<>();
+		String output = executeDiscoveryScript(discoveryScriptFile, gpuAmount, args);
+		if (!output.isEmpty()) {
+			String[] indexes = output.split(",");
+			for (String index : indexes) {
+				if (!StringUtils.isNullOrWhitespaceOnly(index)) {
+					gpuResources.add(new GPUInfo(index.trim()));
+				}
+			}
+		}
+		LOG.info("Discover GPU resources: {}.", gpuResources);
+		return Collections.unmodifiableSet(gpuResources);
+	}
+
+	private String executeDiscoveryScript(File discoveryScript, long gpuAmount, String args) throws Exception {
+		final String cmd = discoveryScript.getAbsolutePath() + " " + gpuAmount + " " + args;
+		final Process process = Runtime.getRuntime().exec(cmd);
```

Comment:
We should add a finally block to the end of the try-with-resources block which create the `BufferedReaders` and make sure that `process` is being destroyed. We can do this by calling `Process.destroyForcibly()`.